protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,628 @@
+        Self { local_peer_id }
+    }
+}
+
+impl IntoProtocolsHandler for Prototype {
+    type Handler = Handler;
+
+    fn into_handler(self, remote_peer_id: &PeerId, endpoint: &ConnectedPoint) -> Self::Handler {
+        Handler {
+            remote_peer_id: *remote_peer_id,
+            remote_addr: endpoint.get_remote_address().clone(),
+            local_peer_id: self.local_peer_id,
+            queued_events: Default::default(),
+            pending_error: Default::default(),
+            reservation: None,
+            alive_lend_out_substreams: Default::default(),
+            circuit_deny_futs: Default::default(),
+            send_error_futs: Default::default(),
+            keep_alive: KeepAlive::Yes,
+        }
+    }
+
+    fn inbound_protocol(&self) -> <Self::Handler as ProtocolsHandler>::InboundProtocol {
+        inbound_stop::Upgrade {}
+    }
+}
+
+pub struct Handler {
+    local_peer_id: PeerId,
+    remote_peer_id: PeerId,
+    remote_addr: Multiaddr,
+    pending_error: Option<
+        ProtocolsHandlerUpgrErr<
+            EitherError<inbound_stop::UpgradeError, outbound_hop::UpgradeError>,
+        >,
+    >,
+    /// Until when to keep the connection alive.
+    keep_alive: KeepAlive,
+
+    /// Queue of events to return when polled.
+    queued_events: VecDeque<
+        ProtocolsHandlerEvent<
+            <Self as ProtocolsHandler>::OutboundProtocol,
+            <Self as ProtocolsHandler>::OutboundOpenInfo,
+            <Self as ProtocolsHandler>::OutEvent,
+            <Self as ProtocolsHandler>::Error,
+        >,
+    >,
+
+    reservation: Option<Reservation>,
+
+    /// Tracks substreams lend out to the transport.
+    ///
+    /// Contains a [`futures::future::Future`] for each lend out substream that
+    /// resolves once the substream is dropped.
+    ///
+    /// Once all substreams are dropped and this handler has no other work,
+    /// [`KeepAlive::Until`] can be set, allowing the connection to be closed
+    /// eventually.
+    alive_lend_out_substreams: FuturesUnordered<oneshot::Receiver<void::Void>>,
+
+    circuit_deny_futs: FuturesUnordered<BoxFuture<'static, (PeerId, Result<(), std::io::Error>)>>,
+
+    send_error_futs: FuturesUnordered<BoxFuture<'static, ()>>,
+}
+
+impl ProtocolsHandler for Handler {
+    type InEvent = In;
+    type OutEvent = Event;
+    type Error = ProtocolsHandlerUpgrErr<
+        EitherError<inbound_stop::UpgradeError, outbound_hop::UpgradeError>,
+    >;
+    type InboundProtocol = inbound_stop::Upgrade;
+    type OutboundProtocol = outbound_hop::Upgrade;
+    type OutboundOpenInfo = OutboundOpenInfo;
+    type InboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(inbound_stop::Upgrade {}, ())
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        inbound_circuit: <Self::InboundProtocol as upgrade::InboundUpgrade<NegotiatedSubstream>>::Output,
```

Comment:
Not sure if this will ever change but that might be clearer.

```suggestion
        inbound_circuit: NegotiatedSubstream,
```

Reply:
I might be missing something. `<Self::InboundProtocol as upgrade::InboundUpgrade<NegotiatedSubstream>>::Output` in this case is a `inbound_stop::Circuit`.

Reply:
Right, my bad! Use `inbound_stop::Circuit` then? :D

Reply:
I would actually argue for the current version. Say we ever rename the `Output` of the `InboundUpgrade`, that change will not propagated beyond the rename itself. Especially useful on larger refactors. I use these _relative_ imports through type parameters instead of concrete types throughout most of the circuit relay v2 implementation.

Do you think the convenience is not worth the type complexity?

Reply:
It makes it quite hard to reason about the code without an IDE because you can't see what it is. But I also get the refactor convenience so happy either way :)

Reply:
Leaving as is then.